full-service/src/service/sync.rs

```diff
@@ -222,172 +222,93 @@ pub fn sync_account_next_chunk(
+                &tx_out,
+                &view_private_key,
+                account_key.as_ref(),
+                conn,
+            );
+
+            received_txos_with_subaddresses_and_key_images.push((
+                block_index,
+                tx_out,
+                amount,
+                subaddress_index,
+                key_image,
+            ));
+        }
 
-            // Match key images to mark existing unspent transactions as spent.
-            let unspent_key_images: MCHashMap<KeyImage, String> =
-                Txo::list_unspent_or_pending_key_images(account_id_hex, None, conn)?;
-            let spent_txos: Vec<(u64, String)> = key_images
-                .into_par_iter()
-                .filter_map(|(block_index, key_image)| {
-                    unspent_key_images
-                        .get(&key_image)
-                        .map(|txo_id_hex| (block_index, txo_id_hex.clone()))
-                })
-                .collect();
-            let num_spent_txos = spent_txos.len();
-            for (block_index, txo_id_hex) in &spent_txos {
-                Txo::update_spent_block_index(txo_id_hex, *block_index, conn)?;
-                TransactionLog::update_pending_associated_with_txo_to_succeeded(
-                    txo_id_hex,
-                    *block_index,
-                    conn,
-                )?;
-            }
+        let num_received_txos = received_txos_with_subaddresses_and_key_images.len();
+
+        // Write received transactions to the database.
+        for (block_index, tx_out, amount, subaddress_index, key_image) in
+            received_txos_with_subaddresses_and_key_images
+        {
+            Txo::create_received(
+                tx_out.clone(),
+                subaddress_index,
+                key_image,
+                amount,
+                block_index,
+                account_id_hex,
+                conn,
+            )?;
+        }
 
-            TransactionLog::update_pending_exceeding_tombstone_block_index_to_failed(
-                &account_id,
-                end_block_index + 1,
+        // Match key images to mark existing unspent transactions as spent.
+        let unspent_key_images: MCHashMap<KeyImage, String> =
+            Txo::list_unspent_or_pending_key_images(account_id_hex, None, conn)?;
+        let spent_txos: Vec<(u64, String)> = key_images
+            .into_par_iter()
+            .filter_map(|(block_index, key_image)| {
+                unspent_key_images
+                    .get(&key_image)
+                    .map(|txo_id_hex| (block_index, txo_id_hex.clone()))
+            })
+            .collect();
+
+        for (block_index, txo_id_hex) in &spent_txos {
+            Txo::update_spent_block_index(txo_id_hex, *block_index, conn)?;
+            TransactionLog::update_pending_associated_with_txo_to_succeeded(
+                txo_id_hex,
+                *block_index,
                 conn,
```

Comment:
The logic and arguments in `update_pending_associated_with_txo_to_succeeded()` will need to be updated to address the issue of multiple transaction logs with the same `txo_id_hex` being marked as succeeded when only one should. 
instead of trying to mirror the change between the two different code branches this function had, I thought it better to consolidate first before applying the fix